scope.go

```diff
@@ -0,0 +1,260 @@
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package dig
+
+import (
+	"bytes"
+	"fmt"
+	"math/rand"
+	"reflect"
+	"sort"
+)
+
+// A ScopeOption modifies the default behavior of Scope; currently,
+// there are no implementations.
+type ScopeOption interface {
+	applyScopeOption(*scopeOptions)
+}
+
+type scopeOptions struct {
+}
+
+// Scope is a scoped DAG of types and their dependencies.
+// A Scope may also have one or more child Scopes that inherit
+// from it.
+type Scope struct {
+	name string
+	// Mapping from key to all the constructor node that can provide a value for that
+	// key.
+	providers map[key][]*constructorNode
+
+	// constructorNodes provided directly to this Scope. i.e. it does not include
+	// any nodes that were provided to the parent Scope this inherited from.
+	nodes []*constructorNode
+
+	// Values that have already been generated directly in the Scope.
+	values map[key]reflect.Value
+
+	// Values groups that have already been generated directly in the Scope.
+	groups map[key][]reflect.Value
+
+	// Source of randomness.
+	rand *rand.Rand
+
+	// Flag indicating whether the graph has been checked for cycles.
+	isVerifiedAcyclic bool
+
+	// Defer acyclic check on provide until Invoke.
+	deferAcyclicVerification bool
+
+	// invokerFn calls a function with arguments provided to Provide or Invoke.
+	invokerFn invokerFn
+
+	// graph of this Scope. Note that this holds the dependency graph of all the
+	// nodes that affect this Scope, not just the ones provided directly to this Scope.
+	gh *graphHolder
+
+	// Parent of this Scope.
+	parentScope *Scope
+
+	// All the child scopes of this Scope.
+	childScopes []*Scope
+}
+
+// Scope creates a new Scope with the given name and options from current Scope.
+// Any constructors that the current Scope knows about, as well as any modifications
+// made to it in the future will be propagated to the child scope.
+// However, no modifications made to the child scope being created will be propagated
+// to the parent Scope.
+func (s *Scope) Scope(name string, opts ...ScopeOption) *Scope {
+	child := &Scope{
+		name:        name,
+		parentScope: s,
+		providers:   make(map[key][]*constructorNode, len(s.providers)),
+		values:      make(map[key]reflect.Value, len(s.values)),
+		groups:      make(map[key][]reflect.Value, len(s.groups)),
+		invokerFn:   s.invokerFn,
+	}
+
+	// child should hold a separate graph holder
+	child.gh = &graphHolder{
+		s:     child,
+		snap:  -1,
+		nodes: make([]*graphNode, len(s.gh.nodes)),
+	}
+
+	for i, graphNode := range s.gh.nodes {
+		child.gh.nodes[i] = graphNode
+	}
+
+	s.childScopes = append(s.childScopes, child)
+	return child
+}
+
+// getScopesUntilRoot creates a list of Scopes
+// have to traverse through until the current node.
```

Comment:
wording: list of Scopes that traverse from current node to the root? Until "the current node" implies top down and doesn't match the function name.